ramips: add support for ipTIME A6004NS-M

antegral opened the pull request:
ipTIME A6004NS-M is a 2.4/5GHz band AC1900 router, based on MediaTek MT7621A.

Specifications:
- SoC: MediaTek MT7621A (880MHz, Duel-Core)
- RAM: DDR3 256MB
- Flash: SPI NOR 16MB (Winbond W25Q128BV)
- WiFi: MediaTek MT7615E (2.4GHz, 5GHz)
- Ethernet: MediaTek MT7530 (WAN x1, LAN x4)
- USB: USB 3.0 x1
- UART: [3.3V, TX, RX, GND] (57600 8N1)

Installation via web interface:
1. Flash initramfs image using OEM's Firmware Update page.
2. Connect to OpenWrt with an SSH connection to `192.168.1.1`.
3. Perform sysupgrade with sysupgrade image.

Revert to stock firmware:
1.  Flash stock firmware via OEM's Recovery mode

How to use OEM's Recovery mode:
1. Power on the device and connect the shell through UART.
2. Connect to the shell and press the `t` key on the keyboard.
3. Set fixed IP with  `192.168.0.2`  with subnet mask  `255.255.255.0`
4.  Flash image via TFTP to  `192.168.0.1`

Additional Notes:
1. The higher the 5Ghz Frequency, the lower the stability. It is recommended to use less than 5.775Ghz.
2. If the 5Ghz frequency is too high, 5Ghz may not work.
 
Signed-off-by: SeongUk Moon <antegral@antegral.net>

mans0n commented:
Hi, it mostly looks good, but the PR description should go into the commit message.
https://openwrt.org/docs/guide-developer/frequent-pr-mistakes#know_the_difference_between_the_pr_pull_request_and_the_commit

mans0n commented on target/linux/ramips/dts/mt7621_iptime_a6004ns-m.dts:
```diff
@@ -0,0 +1,183 @@
+			#size-cells = <1>;
+
+			uboot: partition@0 {
+				label = "u-boot"; 
```
Remove the trailing space.

mans0n commented on target/linux/ramips/dts/mt7621_iptime_a6004ns-m.dts:
```diff
@@ -0,0 +1,183 @@
+
+&state_default {
+	gpio {
+			groups = "i2c", "uart3", "jtag", "wdt";
+			function = "gpio";
```
Wrong indent.

mans0n commented on target/linux/ramips/dts/mt7621_iptime_a6004ns-m.dts:
```diff
@@ -0,0 +1,183 @@
+	macaddr_uboot_1fc40: macaddr@1fc40 {
+		reg = <0x1fc40 0x6>;
+	};
+};
```
Add an empty newline at EOF.

mans0n commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -796,6 +796,17 @@ define Device/iptime_a6ns-m
 endef
 TARGET_DEVICES += iptime_a6ns-m
 
+define Device/iptime_a6004ns-m
```
It sounds strange but:
```sh
$ echo -e "a6ns-m\na6004ns-m" | sort
a6004ns-m
a6ns-m
```
So this should be moved above the a6ns-m.

mans0n reviewed (commented):
Just a few comments:
(AFAIK A6004NS-M is same as A6ns-M except for the RAM size, so this won't be that hard)

mans0n commented:
> Additional Notes:
>     1. The higher the 5Ghz Frequency, the lower the stability. It is recommended to use less than 5.775Ghz.
>     2. If the 5Ghz frequency is too high, 5Ghz may not work.

Could you elaborate? Does it also apply to the stock FW or not?

mans0n commented:
> Ethernet: MediaTek MT7530 (WAN x1, LAN x4)

Is it a separate chip or SoC built-in one?

antegral commented:
> > Additional Notes:
> >     1. The higher the 5Ghz Frequency, the lower the stability. It is recommended to use less than 5.775Ghz.
> >     2. If the 5Ghz frequency is too high, 5Ghz may not work.
> 
> Could you elaborate? Does it also apply to the stock FW or not?

I don't know why, but if you raise the 5Ghz channel above a certain channel, there will be a problem with the 5Ghz radio or it won't work. However, the lower the channel, the more stable it works.

antegral commented:
> > Ethernet: MediaTek MT7530 (WAN x1, LAN x4)
> 
> Is it a separate chip or SoC built-in one?

As a result of decomposing it myself, three parts were identified as heat spreaders. It's hard to proceed anymore, so we don't know exactly, but it's presumed to be an SoC built-in.

antegral commented:
> Hi, it mostly looks good, but the PR description should go into the commit message.
> https://openwrt.org/docs/guide-developer/frequent-pr-mistakes#know_the_difference_between_the_pr_pull_request_and_the_commit

Thank you. I modified it.

mans0n commented:
> I don't know why, but if you raise the 5Ghz channel above a certain channel, there will be a problem with the 5Ghz radio or it won't work. However, the lower the channel, the more stable it works.
> 
> Stock firmware worked normally no matter what channel it chose.

Thanks for your reporting. So you're saying that

* 5785 MHz [157]
* 5805 MHz [161]
* 5825 MHz [165]

channels are not really stable?

> As a result of decomposing it myself, three parts were identified as heat spreaders. It's hard to proceed anymore, so i don't know exactly, but it's presumed to be an SoC built-in.

Then please clarify it in the commit message, as the current description makes confusion. If my guess that a6ns-m is a variant of a6004ns-m with less RAM is correct, then the switch should be SoC built-in indeed. FYI you can verify it by checking whether the PCB patterns from Ethernet ports are directly connected to the SoC or not.

And it's my personal curiosity, could you let me know the result of `iw list`?

mans0n commented:
Come to think of it, we can create a shared dtsi file for a6ns-m and a6004ns-m. Please refer to the WeVO's [dtsi](https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/dts/mt7621_wevo_w2914ns-v2.dtsi) and [dts](https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/dts/mt7621_wevo_w2914ns-v2.dts).

mans0n commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -785,6 +785,17 @@ define Device/iptime_a3004t
+define Device/iptime_a6004ns-m
+  $(Device/dsa-migration)
+  IMAGE_SIZE := 16128k
+  UIMAGE_NAME := a6004nsm
```
Shouldn't it be `a6004nm`? How did you manage to install with this?

antegral replied:
There may be a problem because I wrote the file for personal use, but I rewritten it and uploaded it this time. The image name is a6004nm accurate, and we will revise it soon.

mans0n replied:
I hope you've tested this at runtime, as I won't accept a PR that has not been tested properly.

antegral commented:
> Come to think of it, we can create a shared dtsi file for a6ns-m and a6004ns-m. Please refer to the WeVO's [dtsi](https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/dts/mt7621_wevo_w2914ns-v2.dtsi) and [dts](https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/dts/mt7621_wevo_w2914ns-v2.dts).

a6ns-m and a6004ns-m actually use the same dts code. Then, should i create a dtsi file and import it to the dts of each model?

antegral commented:
> Thanks for your reporting. So you're saying that
> 
> * 5785 MHz [157]
> * 5805 MHz [161]
> * 5825 MHz [165]
> 
> channels are not really stable?

Yes, it is. As the channel goes up, instability increases, and even 5Ghz radio doesn't work. However, lowering the channel again will solve it.

antegral commented:
I will record the execution results of the `iw list` command and the operation results of each channel in detail and show you soon.

antegral commented:
I disassembled the device myself and took a picture of the device at that time.

![KakaoTalk_20220104_205452898_02](https://user-images.githubusercontent.com/60401462/148055482-39f85982-51b1-40e6-99e0-86c5d66d5d44.jpg)

mans0n commented:
Thanks for the log and the picture. Now I'm convinced that two boards are indeed identical.

> a6ns-m and a6004ns-m actually use the same dts code. Then, should I create a dtsi file and import it to the dts of each model?
> If so, do I have to modify the dts code of a6ns-m?

Yes, a6ns-m dts also needs to be changed, that's why we're creating a common dtsi file.

> - WiFi: MediaTek MT7615E (2.4GHz, 5GHz)

This is also confusing. It sounds like there's only one MT7615 chip, but the board should have two separate WLAN chips. Have a look how I described A6ns-M. a972b1fb5f5d

antegral commented:
> Yes, a6ns-m dts also needs to be changed, that's why we're creating a common dtsi file.

Is there any rule for the dtsi file name?

mans0n commented:
> > Yes, a6ns-m dts also needs to be changed, that's why we're creating a common dtsi file.
> 
> Is there any rule for the dtsi file name?

Not really, but I would recommend `mt7621_iptime_a6004ns-m.dtsi`, as a6004ns-m was released before a6ns-m.

antegral commented:
@mans0n By modifying the dts file of a6ns-m, is a separate commitment from the additional commitment message required? If necessary, how can I fill it out?

mans0n commented:
Sorry, has been busy for a while.

It is true that unrelated multiple changes in a single commit is not preferable, but considering 35aecc9d4a73 and c12ef2f76098, I think in this case just one commit would suffice.

antegral committed:
ramips: add support for ipTIME A6004NS-M

ipTIME A6004NS-M is a 2.4/5GHz band AC1900 router, based on MediaTek MT7621A.

Specifications:
- SoC: MediaTek MT7621A (880MHz, Duel-Core)
- RAM: DDR3 256MB
- Flash: SPI NOR 16MB (Winbond W25Q128BV)
- WiFi: MediaTek MT7615E (2.4GHz, 5GHz)
- Ethernet: MediaTek MT7530 (WAN x1, LAN x4, SoC built-in Estimated)
- USB: USB 3.0 x1
- UART: [3.3V, TX, RX, GND] (57600 8N1)

Installation via web interface:
1. Flash initramfs image using OEM's Firmware Update page.
2. Connect to OpenWrt with an SSH connection to `192.168.1.1`.
3. Perform sysupgrade with sysupgrade image.

Revert to stock firmware:
1.  Flash stock firmware via OEM's Recovery mode

How to use OEM's Recovery mode:
1. Power on the device and connect the shell through UART.
2. Connect to the shell and press the `t` key on the keyboard.
3. Set fixed IP with  `192.168.0.2`  with subnet mask  `255.255.255.0`
4. Flash image via TFTP to  `192.168.0.1`

Additional Notes:
1. The higher the 5Ghz Frequency, the lower the stability. It is recommended to use less than 5.775Ghz.
2. If the 5Ghz frequency is too high, 5Ghz may not work.
3. A6ns-M use shared dtsi file of A6004NS-M. (reference: /mt7621_iptime_a6004ns-m.dtsi).

Signed-off-by: SeongUk Moon <antegral@antegral.net>

mans0n commented:
Thanks! Merged your patch as 9c8ccdcdc86c.
